Add support for new LVK HDF injection file format

spxiwh opened the pull request:
This PR adds support for the new LVK HDF injection file format. We do this in the executable that translates from input injection files into the PyCBC format so that we do not make any changes in PyCBC internals.

We also prepare for removal of XML support in the InjectionSet by copying the XML reading stuff into this executable, so that one could remove *all* of the XML code in the InjectionClass at a later date. (For now InjectionSet still supports XML, so this is just *preparing* for this removal, not actually *doing* it).

spxiwh commented:
I note that because of a [necessary] rename here the diff just shows the whole file. If you want to see the actual changes go to:

https://github.com/gwastro/pycbc/compare/9700461..spxiwh:pr_injformat_support

(If you want me to rename the file again, let's do that *last* so that we can still view this diff until that point)

tdent commented:
Is the first test failure ('run small search') caused by something we know about ? 
Also do we need all the FIXMEs ?

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,181 @@
+    """Reads injections from new-style LVK injection files
+    """
+
+    translation_dict = {
```
should this dict rather be defined outside the class?  ie do we gain anything from putting it inside?

spxiwh replied:
The dictionary is a property of the new format, so I'd rather keep all of the "code to handle the new format" within the single class. If we have to support multiple input formats here I think it would keep the code cleaner.

tdent replied:
OK, so it looks like this is a thing called a 'class variable' which is tied to the class, not the instance (`self`). 
I _think_ this requires it to be referenced as `LVKNewStyleInjectionSet.translation_dict`

spxiwh replied:
It is indeed a class variable. These *can* be accessed using `self.XXX`:
```
https://www.atatus.com/blog/class-variables-vs-instance-variables-in-java/#accessing-a-class-variable
```
I think a reason not to is doing something like:
```
self.class_variable = 'BLAH'
```
may not make it clear that it would change for more than self!

Here the class variable is a constant, and so I *think* it's okay, and avoids having to write out the class name every time.

The alternative would be to hardcode these things in `__init__` as an instance variable ...

tdent replied:
OK .. as long as it works as intended

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,181 @@
+        'd_lum': 'distance',
+        'f22_start': 'f_lower',
+        'cbc_model': 'approximant',
+        'longAscNodes': 'FIXME', # FIXME
```
Interpreting the 'FIXME' as a request for guidance : this seems to be `parameter.long_asc_nodes`

spxiwh replied:
Done

tdent commented:
I have checked the 'ESSENTIAL' parameters in the new file spec, the code seems correct for those.

ahnitz commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,181 @@
+    translation_dict = {
+        'mass1_det': 'mass1',
+        'mass2_det': 'mass2',
+        'f22_ref_spin': 'fref', # ????
```
I think this should be 'f_ref'

tdent replied:
Yes.

spxiwh replied:
Oops, typo. Thanks!

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,181 @@
+        'f22_start': 'f_lower',
+        'cbc_model': 'approximant',
+        'longAscNodes': 'FIXME', # FIXME
+        'eccentricity': 'FIXME', # FIXME (check eccentricity params!)
```
I checked : `parameters.eccentricity`

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,181 @@
+        'cbc_model': 'approximant',
+        'longAscNodes': 'FIXME', # FIXME
+        'eccentricity': 'FIXME', # FIXME (check eccentricity params!)
+        'meanPerAno': 'FIXME', # FIXME
```
`parameters.mean_per_ano`

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,181 @@
+        'longAscNodes': 'FIXME', # FIXME
+        'eccentricity': 'FIXME', # FIXME (check eccentricity params!)
+        'meanPerAno': 'FIXME', # FIXME
+        't_co_gps_add': 'FIXME', # FIXME (don't understand why this is needed?)
```
This should be added to t_co_gps, see https://git.ligo.org/waveforms/o4-injection-file/-/merge_requests/3/diffs#diff-content-e34686f17c5639f8fcfbf2f892baa99b42762769

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,181 @@
+        'meanPerAno': 'FIXME', # FIXME
+        't_co_gps_add': 'FIXME', # FIXME (don't understand why this is needed?)
+        'ModeArray': 'FIXME', # FIXME (Will need special handling. Need an example)
+        'ModeArrayJframe': 'FIXME', # FIXME (Does this do anything useful?)
```
Same explanation here. This may or may not be 'useful' but it corresponds to a LALSimInspiral parameters slot, so someone at some point thought it might be.

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,181 @@
+    translation_dict = {
+        'mass1_det': 'mass1',
+        'mass2_det': 'mass2',
+        'f22_ref_spin': 'fref', # ????
```
Yes, `f_ref`. It is called this in the new spec because it's the reference frequency of the 2,2 GW mode at which spins are specified.

spxiwh commented:
The tc issue should be fixed now as well. I think this addresses the comments, but I do want to see if I can reproduce/understand the search workflow failure

spxiwh commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,192 @@
+    def columns_in_pycbc_format(self, subdir):
+        # Loop over columns in input file
+                         
+        for field_name in self.inj_file['cbc_waveform_params'].keys():
```
Should be fixed now. (Also should have fixed the workflow error)

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,194 @@
+        'ModeArrayJframe': 'FIXME', # FIXME (Does this do anything useful?)
+    }
+
+    subdir = 'cbc_waveform_params'
```
Similarly here, if this is not `self.subdir` it ought to be `LVKNewStyleInjectionSet.subdir`, but that will make the code super unreadable ..

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,194 @@
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+""" This program converts standard injection file formats (e.g. XML format
+or the LIGO HDF format) into a PyCBC hdf format that includes a template hash
```
Is this still true (ie there is a template hash in injection files ??)

spxiwh replied:
It never was true. I think this was copied from the template bank converter without updating. Fixed.

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,194 @@
+        for field_name in self.inj_file[self.subdir].keys():
+            if field_name in self.translation_dict:
+                if field_name == 't_co_gps_add':
+                    continue
```
The handling of this special case is kind of split up between this function and the pack_data function, which I think is too complicated

spxiwh replied:
Changed as suggested.

tdent commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,194 @@
+    def pack_data_into_pycbc_format_input(self):
+        data = {}
+        for lvk_name, pycbc_name in self.columns_in_pycbc_format():
+            if lvk_name == 't_co_gps':
```
can't we also have `if lvk_name == 't_co_gps_add': continue` ?

spxiwh replied:
Changed as suggested.

spxiwh commented on bin/pycbc_convertinjfiletohdf:
```diff
@@ -0,0 +1,194 @@
+        'longAscNodes': 'long_asc_nodes',
+        'eccentricity': 'eccentricity',
+        'meanPerAno': 'mean_per_ano',
+        't_co_gps_add': None,
```
I want it to be clear in the code that we are handling this column. So even though we could remove it , I do want to explicitly include it here. Also, if anyone does change the code at a later date, the code will then fail if trying to add this column, rather than copying a t_co_gps_add entry across, which would be ignored.
